### PR TITLE
【KernelGen】Add concatenate operator

### PR DESCRIPTION
**Pull request by Schopenhauer-loves-Hegel**
### PR Category
Operator

### Type of Change
New Feature

### Description
Add `concatenate` operator implementation with Triton kernel.

- Implementation mode: `manual_kernel`
- Accuracy test: 136/136 passed

### Issue
N/A

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [ ] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### Performance
**torch.bfloat16**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [512, 512] x3, dim=0 | 0.0128 | 0.0095 | 1.343 |
| [512, 2048] x3, dim=0 | 0.0239 | 0.0158 | 1.513 |
| [512, 2048] x3, dim=-1 | 0.0237 | 0.0157 | 1.506 |

**torch.float16**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [64, 64] x3, dim=0 | 0.0106 | 0.0094 | 1.126 |
| [64, 64] x3, dim=-1 | 0.0102 | 0.0081 | 1.265 |
| [256, 256] x3, dim=0 | 0.0098 | 0.0081 | 1.213 |
| [256, 256] x3, dim=-1 | 0.0091 | 0.0076 | 1.199 |
| [512, 512] x3, dim=0 | 0.0120 | 0.0107 | 1.129 |
| [512, 2048] x3, dim=0 | 0.0228 | 0.0158 | 1.442 |
| [1024, 512] x3, dim=-1 | 0.0168 | 0.0115 | 1.461 |

**torch.float32**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [64, 64] x3, dim=0 | 0.0084 | 0.0070 | 1.186 |
| [256, 256] x3, dim=0 | 0.0101 | 0.0085 | 1.189 |
| [64, 64, 16] x3, dim=-1 | 0.0109 | 0.0084 | 1.292 |

**torch.int16**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [512, 512] x3, dim=0 | 0.0130 | 0.0095 | 1.370 |
| [512, 2048] x3, dim=0 | 0.0229 | 0.0158 | 1.452 |

**torch.int32**

| Shape | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|-------------------:|-------------------:|--------:|
| [64, 64] x3, dim=0 | 0.0090 | 0.0070 | 1.273 |
| [64, 64, 1] x3, dim=0 | 0.0093 | 0.0070 | 1.318 |

**Overall: median speedup = 1.292x, mean speedup = 1.310x** (17 data points)

---
_Generated by auto_gen tool with Claude Code_

### CODE REVIEW

**Comment by Schopenhauer-loves-Hegel** [comment]
Closing in favor of grouped PRs to reduce review overhead.

**Comment by factnn** [comment]
## concatenate Benchmark Speedup Results

### float16
| Size | dim | Torch (ms) | Gems (ms) | Speedup |
|------|-----|-----------|----------|---------|
| [64,64]x3 | 0 | 0.010 | 0.008 | 1.180 |
| [256,256]x3 | 0 | 0.009 | 0.009 | 0.957 |
| [512,512]x3 | 0 | 0.013 | 0.010 | 1.360 |
| [512,1024]x3 | 0 | 0.013 | 0.011 | 1.154 |
| [512,2048]x3 | 0 | 0.019 | 0.016 | 1.202 |
| [1024,512]x3 | -1 | 0.013 | 0.012 | 1.150 |

### float32
| Size | dim | Torch (ms) | Gems (ms) | Speedup |
|------|-----|-----------|----------|---------|
| [512,512]x3 | 0 | 0.016 | 0.013 | 1.231 |
| [512,1024]x3 | 0 | 0.019 | 0.016 | 1.188 |
| [512,2048]x3 | 0 | 0.027 | 0.026 | 1.038 |
| [1024,512]x3 | 0 | 0.019 | 0.017 | 1.118 |

### bfloat16
| Size | dim | Torch (ms) | Gems (ms) | Speedup |
|------|-----|-----------|----------|---------|
| [512,512]x3 | 0 | 0.013 | 0.010 | 1.300 |
| [512,1024]x3 | 0 | 0.013 | 0.012 | 1.083 |
| [512,2048]x3 | 0 | 0.019 | 0.016 | 1.188 |
| [1024,512]x3 | -1 | 0.013 | 0.012 | 1.083 |

### int32
| Size | dim | Torch (ms) | Gems (ms) | Speedup |
|------|-----|-----------|----------|---------|
| [512,512]x3 | 0 | 0.014 | 0.012 | 1.200 |
| [512,1024]x3 | 0 | 0.018 | 0.016 | 1.144 |
| [1024,32]x3 | 0 | 0.012 | 0.007 | 1.543 |

File: tests/test_concatenate.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                    dim - rank,
+                ]
+            )
+    return results
+
+
+@pytest.mark.concatenate
+@pytest.mark.parametrize("shape, dim", gen_cat_shapes_dim(CAT_SHAPES))
+@pytest.mark.parametrize("dtype", FLOAT_DTYPES + INT_DTYPES)
+def test_accuracy_concatenate(shape, dim, dtype):
```
Comment (tengqm) [inline tests/test_concatenate.py]:
```suggestion
def test_concatenate(shape, dim, dtype):
```

File: tests/test_concatenate.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        (((0, 3), (2, 3)), 0),
+        (((0, 3), (0, 3)), 0),
+        (((0,), (0,)), 0),
+        (((0,), (1, 3)), -1),
+        (((0,), (1, 2, 3)), -2),
+        (((0,), (1, 1, 2, 3)), -3),
+    ],
+)
+@pytest.mark.parametrize("dtype", [torch.float32])
+def test_accuracy_concatenate_empty_tensor(shape, dim, dtype):
```
Comment (tengqm) [inline tests/test_concatenate.py]:
```suggestion
def test_concatenate_empty_tensor(shape, dim, dtype):
```

**Review by tengqm — COMMENTED** [review]
Overall LGTM.

**Review by tengqm — DISMISSED** [review]

**Comment by factnn** [comment]
CI python-op failure is not caused by this PR. `concatenate` is a thin wrapper that directly calls the existing `cat` operator (`src/flag_gems/ops/concatenate.py` → `from flag_gems.ops.cat import cat`). The Triton compilation error originates from the upstream `cat` kernel's `if/elif` dynamic pointer assignment, which is a pre-existing issue.

Also fixed isort formatting in `benchmark/test_concatenate_perf.py`.

File: benchmark/test_concatenate_perf.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+from typing import Generator
+
+import pytest
+import torch
+
+from benchmark.attri_util import FLOAT_DTYPES, INT_DTYPES, BenchLevel
+from benchmark.performance_utils import Benchmark, Config
```
Comment (tengqm) [inline benchmark/test_concatenate_perf.py]: Rebase to master please

**Comment by factnn** [comment]
## UT Results

Branch rebased to upstream/master and tests passed:

```
python -m pytest tests/ -m concatenate --ref cpu -x -q
136 passed
```

**Review by bin913 — APPROVED** [review]
